File: ruby/0056-merge-intervals.rb
def merge(intervals)
    intervals.sort!

    output = [intervals[0]]

    (1..intervals.size()-1).each do |i|
        last_end = output[-1][1]

        if (last_end>= intervals[i][0])
            output[-1][1] = [last_end, intervals[i][1]].max
        else
            output.append(intervals[i])
        end
    end

    return output

end

File: typescript/0056-merge-intervals.ts
function merge(intervals: number[][]): number[][] {
    intervals.sort(([aStart, aEnd], [bStart, bEnd]) => aStart !== bStart ? aStart - bStart : aEnd - bEnd);
    
    return mergeInterval(intervals);
};

function mergeInterval(intervals: number[][]): number[][] {
    let merged = [];
    let prev = intervals.shift();
    
    for (const curr of intervals) {
        const [prevStart, prevEnd] = prev;
        const [currStart, currEnd] = curr;
        
        // Overlap occurs
        if (currStart <= prevEnd) {
            prev[1] = Math.max(prev[1], curr[1]);
            continue;
        }
        
        merged.push(prev);
        prev = curr;
    }
    return [...merged, prev];
}